# Optimizing the Python Code for Big Data 
Balancing Coding Complexity against Computational Complexity 

    
    AUTHOR: Dr. Roy Jafari 

# Chapter 5: Picking up the right tool 

## Challenge 2: Restructuring and Reformulating Data 

In this challenge, we will use a large dataset, `US_Shops_simulated.csv`, which contains about 5 million rows. Due to the size of this data, selecting the wrong tool can make simple restructuring and reformulating tasks take significantly longer. Use the following steps to complete this task.

1. The following code uses `pd.read_csv()` to read `US_Shops_simulated.csv` into `order_df`. Run the code and study the dataset. Specify columns that are providing an index versus the columns that are presenting values.

```
import pandas as pd
order_df = pd.read_csv('US_Shops_simulated.csv')
order_df
```

In [2]:
import pandas as pd
order_df = pd.read_csv('US_Shops_simulated.csv')
order_df

Date  Month  Year    State Location  number_of_customer_visits  \
0        2022-09-30      9  2022  Alabama    Shop1                        499   
1        2022-09-30      9  2022  Alabama    Shop2                        504   
2        2022-09-30      9  2022  Alabama    Shop3                        759   
3        2022-09-30      9  2022  Alabama    Shop4                        715   
4        2022-09-30      9  2022  Alabama    Shop5                        538   
...             ...    ...   ...      ...      ...                        ...   
5470615  2019-10-02     10  2019  Wyoming  Shop119                        476   
5470616  2019-10-02     10  2019  Wyoming  Shop120                        811   
5470617  2019-10-02     10  2019  Wyoming  Shop121                        637   
5470618  2019-10-02     10  2019  Wyoming  Shop122                        612   
5470619  2019-10-02     10  2019  Wyoming  Shop123                        215   

         revenue  profit  
0        1377.74   27.55  
1        3733.13   63.46  
2        6155.49   98.49  
3        7390.96  369.55  
4        6502.27  279.60  
...          ...     ...  
5470615  6968.16  222.98  
5470616  6095.48  176.77  
5470617  1818.64   47.28  
5470618  7336.66  278.79  
5470619  2419.83   38.72  

[5470620 rows x 8 columns]

**Answer**: `Month` and `Year` are redundant in the presence of `Date`. `Date`, `State`, and `Location` are the index columns, while `number_of_customer_visits`, `revenue`, and `profit` are the value columns.

2. The following code uses the `.groupby()` and `.size()` functions to get all the indices for all the shops in this dataset. Run the following code and study its output.

```
shop_index = order_df.groupby(['State', 'Location']).size().index
print(shop_index)
```

In [3]:
shop_index = order_df.groupby(['State','Location']).size().index
print(shop_index)

MultiIndex([('Alabama',  'Shop1'),
            ('Alabama', 'Shop10'),
            ('Alabama', 'Shop11'),
            ('Alabama', 'Shop12'),
            ('Alabama', 'Shop13'),
            ('Alabama', 'Shop14'),
            ('Alabama', 'Shop15'),
            ('Alabama', 'Shop16'),
            ('Alabama', 'Shop17'),
            ('Alabama', 'Shop18'),
            ...
            ('Wyoming', 'Shop90'),
            ('Wyoming', 'Shop91'),
            ('Wyoming', 'Shop92'),
            ('Wyoming', 'Shop93'),
            ('Wyoming', 'Shop94'),
            ('Wyoming', 'Shop95'),
            ('Wyoming', 'Shop96'),
            ('Wyoming', 'Shop97'),
            ('Wyoming', 'Shop98'),
            ('Wyoming', 'Shop99')],
           names=['State', 'Location'], length=4996)


3. Similar to step 2, the following code uses the `.groupby()` and `.size()` functions to get all the indices for all the dates in this dataset. Run the following code and study its output.

```
date_index = order_df.groupby(['Date']).size().index
print(date_index)
```


In [5]:
date_index = order_df.groupby(['Date']).size().index
print(date_index)

Index(['2019-10-02', '2019-10-03', '2019-10-04', '2019-10-05', '2019-10-06',
       '2019-10-07', '2019-10-08', '2019-10-09', '2019-10-10', '2019-10-11',
       ...
       '2022-09-21', '2022-09-22', '2022-09-23', '2022-09-24', '2022-09-25',
       '2022-09-26', '2022-09-27', '2022-09-28', '2022-09-29', '2022-09-30'],
      dtype='object', name='Date', length=1095)


4. The following code uses `shop_index` and `date_index` to create the new table, `rev_shop_df`, to contain the daily revenue of all the shops for all the dates from `order_df`. In this table, each row represents a shop and the columns represent the dates.

```
rev_shop_df = pd.DataFrame(index=shop_index, columns=date_index)
rev_shop_df
```

In [8]:
rev_shop_df = pd.DataFrame(index = shop_index,
                      columns = date_index )
rev_shop_df

Date             2019-10-02 2019-10-03 2019-10-04 2019-10-05 2019-10-06  \
State   Location                                                          
Alabama Shop1           NaN        NaN        NaN        NaN        NaN   
        Shop10          NaN        NaN        NaN        NaN        NaN   
        Shop11          NaN        NaN        NaN        NaN        NaN   
        Shop12          NaN        NaN        NaN        NaN        NaN   
        Shop13          NaN        NaN        NaN        NaN        NaN   
...                     ...        ...        ...        ...        ...   
Wyoming Shop95          NaN        NaN        NaN        NaN        NaN   
        Shop96          NaN        NaN        NaN        NaN        NaN   
        Shop97          NaN        NaN        NaN        NaN        NaN   
        Shop98          NaN        NaN        NaN        NaN        NaN   
        Shop99          NaN        NaN        NaN        NaN        NaN   

Date             2019-10-07 2019-10-08 2019-10-09 2019-10-10 2019-10-11  ...  \
State   Location                                                         ...   
Alabama Shop1           NaN        NaN        NaN        NaN        NaN  ...   
        Shop10          NaN        NaN        NaN        NaN        NaN  ...   
        Shop11          NaN        NaN        NaN        NaN        NaN  ...   
        Shop12          NaN        NaN        NaN        NaN        NaN  ...   
        Shop13          NaN        NaN        NaN        NaN        NaN  ...   
...                     ...        ...        ...        ...        ...  ...   
Wyoming Shop95          NaN        NaN        NaN        NaN        NaN  ...   
        Shop96          NaN        NaN        NaN        NaN        NaN  ...   
        Shop97          NaN        NaN        NaN        NaN        NaN  ...   
        Shop98          NaN        NaN        NaN        NaN        NaN  ...   
        Shop99          NaN        NaN        NaN        NaN        NaN  ...   

Date             2022-09-21 2022-09-22 2022-09-23 2022-09-24 2022-09-25  \
State   Location                                                          
Alabama Shop1           NaN        NaN        NaN        NaN        NaN   
        Shop10          NaN        NaN        NaN        NaN        NaN   
        Shop11          NaN        NaN        NaN        NaN        NaN   
        Shop12          NaN        NaN        NaN        NaN        NaN   
        Shop13          NaN        NaN        NaN        NaN        NaN   
...                     ...        ...        ...        ...        ...   
Wyoming Shop95          NaN        NaN        NaN        NaN        NaN   
        Shop96          NaN        NaN        NaN        NaN        NaN   
        Shop97          NaN        NaN        NaN        NaN        NaN   
        Shop98          NaN        NaN        NaN        NaN        NaN   
        Shop99          NaN        NaN        NaN        NaN        NaN   

Date             2022-09-26 2022-09-27 2022-09-28 2022-09-29 2022-09-30  
State   Location                                                         
Alabama Shop1           NaN        NaN        NaN        NaN        NaN  
        Shop10          NaN        NaN        NaN        NaN        NaN  
        Shop11          NaN        NaN        NaN        NaN        NaN  
        Shop12          NaN        NaN        NaN        NaN        NaN  
        Shop13          NaN        NaN        NaN        NaN        NaN  
...                     ...        ...        ...        ...        ...  
Wyoming Shop95          NaN        NaN        NaN        NaN        NaN  
        Shop96          NaN        NaN        NaN        NaN        NaN  
        Shop97          NaN        NaN        NaN        NaN        NaN  
        Shop98          NaN        NaN        NaN        NaN        NaN  
        Shop99          NaN        NaN        NaN        NaN        NaN  

[4996 rows x 1095 columns]

5. The following code uses a for loop and the `.query()` function to wrangle `order_df` into `rev_shop_df`. Note that `%%time` is used to record how long the code will take to run. Run the following code and note its outputs.

```
%%time
for state,location in shop_index:
    wdf = (order_df
           .query(f'State == "{state}"')
           .query(f'Location == "{location}"')
           .copy()
    )
    rev_shop_df.loc[state,location] = (
        wdf.set_index('Date').number_of_customer_visits)
rev_shop_df
```

In [119]:
%%time
for state,location in shop_index:
    wdf = (order_df
           .query(f'State == "{state}"')
           .query(f'Location == "{location}"')
           .copy()
    )
    rev_shop_df.loc[state,location] = (
        wdf.set_index('Date').number_of_customer_visits)
rev_shop_df

CPU times: user 3min 43s, sys: 17.2 s, total: 4min
Wall time: 4min 10s


Date             2019-10-02 2019-10-03 2019-10-04 2019-10-05 2019-10-06  \
State   Location                                                          
Alabama Shop1           837        519        233        478        549   
        Shop10          308        133        155        273        373   
        Shop11          837        454        544        162        432   
        Shop12          154        464        704        361        952   
        Shop13          361        155        806        873        469   
...                     ...        ...        ...        ...        ...   
Wyoming Shop95          302        513        865        195        675   
        Shop96          101        391        987        755        287   
        Shop97          975        285        577        852        738   
        Shop98          240        782        168        727        224   
        Shop99          943        127        377        884        682   

Date             2019-10-07 2019-10-08 2019-10-09 2019-10-10 2019-10-11  ...  \
State   Location                                                         ...   
Alabama Shop1           363        351        262        701        352  ...   
        Shop10          158        690        359        859        344  ...   
        Shop11          733        249        282        890        587  ...   
        Shop12          421        665        500        219        169  ...   
        Shop13          224        178        868        568        269  ...   
...                     ...        ...        ...        ...        ...  ...   
Wyoming Shop95          977        971        413        943        973  ...   
        Shop96          380        693        534        840        957  ...   
        Shop97          672        664        589        730        978  ...   
        Shop98          841        224        576        638        412  ...   
        Shop99          715        330        433        494        388  ...   

Date             2022-09-21 2022-09-22 2022-09-23 2022-09-24 2022-09-25  \
State   Location                                                          
Alabama Shop1           277        544        347        666        734   
        Shop10          836        744        461        206        424   
        Shop11          689        714        202        282        193   
        Shop12          148        818        654        121        867   
        Shop13          535        893        837        411        490   
...                     ...        ...        ...        ...        ...   
Wyoming Shop95          572        542        274        309        917   
        Shop96          191        525        373        576        987   
        Shop97          960        703        437        241        578   
        Shop98          296        672        814        437        792   
        Shop99          249        975        467        130        680   

Date             2022-09-26 2022-09-27 2022-09-28 2022-09-29 2022-09-30  
State   Location                                                         
Alabama Shop1           368        340        374        418        499  
        Shop10          488        791        100        844        983  
        Shop11          256        887        365        127        924  
        Shop12          400        439        647        485        634  
        Shop13          964        690        543        100        450  
...                     ...        ...        ...        ...        ...  
Wyoming Shop95          493        349        132        378        138  
        Shop96          316        949        191        411        953  
        Shop97          348        864        619        548        578  
        Shop98          305        449        882        308        205  
        Shop99          171        375        206        851        331  

[4996 rows x 1095 columns]

**Answer:** It took my computer 4 minutes and 3 seconds, or 243 seconds, to complete the task using the for loop and `.query()` approach. 

6. The following code performs an even bigger task than the one done in steps 4 and 5, but with more appropriate tools. While the task in steps 4 and 5 was to wrangle the revenue data, this step also wrangles the `profit` and `number_of_customer_visits` data. In this code, the functions `.drop()`, `.set_index()`, and `.unstack()` are employed. Note that the critical function enabling this restructuring is `.unstack()`.

```
%%time
shop_df = (
    order_df
    .drop(columns=['Month', 'Year'])
    .set_index(['State', 'Location', 'Date'])
    .unstack()
)
shop_df
```
Run `shop_df['revenue']` to see what we created in steps 4 and 5.

In [13]:
%%time
shop_df = (
    order_df
    .drop(columns=['Month', 'Year'])
    .set_index(['State', 'Location', 'Date'])
    .unstack()
)
shop_df

CPU times: user 1.25 s, sys: 166 ms, total: 1.41 s
Wall time: 1.44 s


number_of_customer_visits                                   \
Date                            2019-10-02 2019-10-03 2019-10-04 2019-10-05   
State   Location                                                              
Alabama Shop1                          837        519        233        478   
        Shop10                         308        133        155        273   
        Shop11                         837        454        544        162   
        Shop12                         154        464        704        361   
        Shop13                         361        155        806        873   
...                                    ...        ...        ...        ...   
Wyoming Shop95                         302        513        865        195   
        Shop96                         101        391        987        755   
        Shop97                         975        285        577        852   
        Shop98                         240        782        168        727   
        Shop99                         943        127        377        884   

                                                                         \
Date             2019-10-06 2019-10-07 2019-10-08 2019-10-09 2019-10-10   
State   Location                                                          
Alabama Shop1           549        363        351        262        701   
        Shop10          373        158        690        359        859   
        Shop11          432        733        249        282        890   
        Shop12          952        421        665        500        219   
        Shop13          469        224        178        868        568   
...                     ...        ...        ...        ...        ...   
Wyoming Shop95          675        977        971        413        943   
        Shop96          287        380        693        534        840   
        Shop97          738        672        664        589        730   
        Shop98          224        841        224        576        638   
        Shop99          682        715        330        433        494   

                             ...     profit                                   \
Date             2019-10-11  ... 2022-09-21 2022-09-22 2022-09-23 2022-09-24   
State   Location             ...                                               
Alabama Shop1           352  ...       8.87     255.61     115.84     220.68   
        Shop10          344  ...      96.02      80.16     215.60      14.63   
        Shop11          587  ...      62.20      46.36      21.32      94.82   
        Shop12          169  ...      12.04     187.34     197.65      83.18   
        Shop13          269  ...      46.00     291.97     460.52      93.61   
...                     ...  ...        ...        ...        ...        ...   
Wyoming Shop95          973  ...     181.28      70.13      23.35     148.37   
        Shop96          957  ...      49.95      42.23     262.93     316.60   
        Shop97          978  ...     570.96     136.14      40.29     115.45   
        Shop98          412  ...       6.02      79.85     180.08     172.95   
        Shop99          388  ...      13.89     137.50     292.02      18.89   

                                                                         \
Date             2022-09-25 2022-09-26 2022-09-27 2022-09-28 2022-09-29   
State   Location                                                          
Alabama Shop1        151.33     156.96      70.63      18.77      69.95   
        Shop10        33.89      11.36     335.82       2.77     445.96   
        Shop11        20.06      89.73     353.26      62.86      19.36   
        Shop12       197.14     266.99     207.88     264.04     237.32   
        Shop13        72.41     310.95     102.98      87.58       3.44   
...                     ...        ...        ...        ...        ...   
Wyoming Shop95       155.46     314.57      50.81      19.26   

**Answer:** It took my computer 1.44 seconds to complete the task using the `.unstack()` approach.

In [12]:
shop_df['revenue']

Date              2019-10-02  2019-10-03  2019-10-04  2019-10-05  2019-10-06  \
State   Location                                                               
Alabama Shop1       10378.80     6346.85     1204.84     6899.45     2163.06   
        Shop10       1896.97      397.00     1877.21     3001.09     3952.68   
        Shop11       3646.81     1946.75     1964.38     1740.04     2149.63   
        Shop12       1908.83     1126.59     8712.70     3300.98    11177.43   
        Shop13       4452.21     1742.20     4593.39     2180.75     2592.16   
...                      ...         ...         ...         ...         ...   
Wyoming Shop95       2671.19     2681.45     3456.54      502.71     6452.32   
        Shop96       1049.79      926.28     3973.66     1882.22     2537.08   
        Shop97       8952.45     4154.16     8215.33     3484.68     4236.86   
        Shop98       1036.80     2446.10     2186.86     2817.85     1277.02   
        Shop99       8657.68      557.53     3954.73     7725.28     9554.82   

Date              2019-10-07  2019-10-08  2019-10-09  2019-10-10  2019-10-11  \
State   Location                                                               
Alabama Shop1        3134.50     1768.34     2739.47     2471.73     1856.10   
        Shop10       1854.13     1524.21     4192.04     2199.04     4958.76   
        Shop11       2810.32     2596.82     4204.62    10447.71     4120.74   
        Shop12       5210.72     9242.17     1205.50      515.74     1105.43   
        Shop13       2495.14     1810.44    10814.41     7714.01     2393.56   
...                      ...         ...         ...         ...         ...   
Wyoming Shop95       6510.73    12552.12     4813.93     7445.93    12080.77   
        Shop96       3999.50     4135.82     4198.31     4058.04    11371.07   
        Shop97       6902.11     6368.42     3002.72     8104.46     4701.25   
        Shop98       3861.87     2040.64     2619.07     6334.06     2765.34   
        Shop99       3996.14     3780.15     4997.69     2996.11     3003.51   

Date              ...  2022-09-21  2022-09-22  2022-09-23  2022-09-24  \
State   Location  ...                                                   
Alabama Shop1     ...     1108.28     6554.11     5036.36     5658.34   
        Shop10    ...     6858.54     5725.82     4067.86      665.17   
        Shop11    ...     2006.37     1717.17     1184.53     3269.79   
        Shop12    ...      547.16     4930.09     5490.33     1697.63   
        Shop13    ...     4181.56     5508.92    12118.92     2400.24   
...               ...         ...         ...         ...         ...   
Wyoming Shop95    ...     4421.56     3506.74     2918.92     3225.34   
        Shop96    ...     1611.28     6032.78     5056.39     6331.97   
        Shop97    ...    13925.76     2959.63     2370.29     2509.77   
        Shop98    ...      753.02     7258.94     5457.06     4434.68   
        Shop99    ...     2777.60     4296.82     6791.11      651.30   

Date              2022-09-25  2022-09-26  2022-09-27  2022-09-28  2022-09-29  \
State   Location                                                               
Alabama Shop1        8901.95     5412.54     5045.26      893.86     1295.38   
        Shop10       2607.18     1032.61     8610.83      307.90     8108.31   
        Shop11        514.34     2039.30     9812.88     2417.76     1613.54   
        Shop12       4285.58     5037.60     5330.34     5867.64     7191.58   
        Shop13       3016.93     6478.08     6057.51     2575.99      214.70   
...                      ...         ...         ...         ...         ...   
Wyoming Shop95       6759.21     7315.63     1104.58     1604.99     1722.55   
        Shop96       5265.64     3602.08     3622.33     1986.21     2331.19   
        Shop97       8327.25     4190.96     9570.53     5867.50     8006.83   
        Shop98       8491.03     2223.14     2566.93     5962.32     3443.75   
        

7. Calculate roughly how much faster the application of `.unstack()` was compared to using the for loop and `.query()`. What is the reason that `.unstack()` is much faster?

**Answer:** The for loop and `.query()` approach took 243 seconds, while the `.unstack()` approach took 1.44 seconds. To calculate the speed improvement:

243\1.44 = 168.75 

The reason `.unstack()` is much faster is that it is optimized for reshaping and pivoting data frames. It performs operations at a lower level in the C code of the pandas library, reducing the overhead associated with Python loops and multiple queries. `.unstack()` leverages efficient memory management and vectorized operations, making it significantly faster than iterating through rows and performing multiple `.query()` calls.

8. The function `.unstack()` has a peer function, `.stack()`. For practice, use `.stack()` to revert `shop_df` back to `order_df`.

**Answer:**

%%time
reverted_order_df = shop_df.stack(future_stack = True).reset_index()
reverted_order_df

9. The set of functions `.stack()` and `.unstack()` are powerful. However, the same tasks can be done using `.melt()` and `.pivot()` functions. To practice, first create `shop_df` using one of these two functions, and revert it back to `order_df` using the other. Pay attention—doing it with `.melt()` might be tricky due to the fact that this function cannot deal with multi-level columns.

**Answer:**

In [113]:
%%time
shop_df = (order_df
 .pivot(
     index=['State', 'Location'], 
     columns='Date', 
     values=['number_of_customer_visits', 'profit', 'revenue'])
)
shop_df

CPU times: user 1.04 s, sys: 121 ms, total: 1.16 s
Wall time: 1.18 s


number_of_customer_visits                                   \
Date                            2019-10-02 2019-10-03 2019-10-04 2019-10-05   
State   Location                                                              
Alabama Shop1                        837.0      519.0      233.0      478.0   
        Shop10                       308.0      133.0      155.0      273.0   
        Shop11                       837.0      454.0      544.0      162.0   
        Shop12                       154.0      464.0      704.0      361.0   
        Shop13                       361.0      155.0      806.0      873.0   
...                                    ...        ...        ...        ...   
Wyoming Shop95                       302.0      513.0      865.0      195.0   
        Shop96                       101.0      391.0      987.0      755.0   
        Shop97                       975.0      285.0      577.0      852.0   
        Shop98                       240.0      782.0      168.0      727.0   
        Shop99                       943.0      127.0      377.0      884.0   

                                                                         \
Date             2019-10-06 2019-10-07 2019-10-08 2019-10-09 2019-10-10   
State   Location                                                          
Alabama Shop1         549.0      363.0      351.0      262.0      701.0   
        Shop10        373.0      158.0      690.0      359.0      859.0   
        Shop11        432.0      733.0      249.0      282.0      890.0   
        Shop12        952.0      421.0      665.0      500.0      219.0   
        Shop13        469.0      224.0      178.0      868.0      568.0   
...                     ...        ...        ...        ...        ...   
Wyoming Shop95        675.0      977.0      971.0      413.0      943.0   
        Shop96        287.0      380.0      693.0      534.0      840.0   
        Shop97        738.0      672.0      664.0      589.0      730.0   
        Shop98        224.0      841.0      224.0      576.0      638.0   
        Shop99        682.0      715.0      330.0      433.0      494.0   

                             ...    revenue                                   \
Date             2019-10-11  ... 2022-09-21 2022-09-22 2022-09-23 2022-09-24   
State   Location             ...                                               
Alabama Shop1         352.0  ...    1108.28    6554.11    5036.36    5658.34   
        Shop10        344.0  ...    6858.54    5725.82    4067.86     665.17   
        Shop11        587.0  ...    2006.37    1717.17    1184.53    3269.79   
        Shop12        169.0  ...     547.16    4930.09    5490.33    1697.63   
        Shop13        269.0  ...    4181.56    5508.92   12118.92    2400.24   
...                     ...  ...        ...        ...        ...        ...   
Wyoming Shop95        973.0  ...    4421.56    3506.74    2918.92    3225.34   
        Shop96        957.0  ...    1611.28    6032.78    5056.39    6331.97   
        Shop97        978.0  ...   13925.76    2959.63    2370.29    2509.77   
        Shop98        412.0  ...     753.02    7258.94    5457.06    4434.68   
        Shop99        388.0  ...    2777.60    4296.82    6791.11     651.30   

                                                                         \
Date             2022-09-25 2022-09-26 2022-09-27 2022-09-28 2022-09-29   
State   Location                                                          
Alabama Shop1       8901.95    5412.54    5045.26     893.86    1295.38   
        Shop10      2607.18    1032.61    8610.83     307.90    8108.31   
        Shop11       514.34    2039.30    9812.88    2417.76    1613.54   
        Shop12      4285.58    5037.60    5330.34    5867.64    7191.58   
        Shop13      3016.93    6478.08    6057.51    2575.99     214.70   
...                     ...        ...        ...        ...        ...   
Wyoming Shop95      6759.21    7315.63    1104.58    1604.99   

In [117]:
%%time
shop_df.columns = shop_df.columns.tolist()
stage_df = shop_df.reset_index().melt(id_vars=['State', 'Location'], value_vars=shop_df.columns.tolist())
stage_df['Value_Type'] = stage_df.variable.apply(lambda sr:sr[0])
stage_df['Date'] = stage_df.variable.apply(lambda sr:sr[1])
reverted_order_df =stage_df.drop(columns = ['variable']).pivot(
     index=['State', 'Location','Date'], 
     columns = 'Value_Type',
     values = 'value').reset_index()
reverted_order_df

CPU times: user 7.76 s, sys: 588 ms, total: 8.35 s
Wall time: 8.48 s


Value_Type    State Location        Date  number_of_customer_visits  profit  \
0           Alabama    Shop1  2019-10-02                      837.0  394.39   
1           Alabama    Shop1  2019-10-03                      519.0  298.30   
2           Alabama    Shop1  2019-10-04                      233.0   66.27   
3           Alabama    Shop1  2019-10-05                      478.0  365.67   
4           Alabama    Shop1  2019-10-06                      549.0   77.87   
...             ...      ...         ...                        ...     ...   
5470615     Wyoming   Shop99  2022-09-26                      171.0   80.35   
5470616     Wyoming   Shop99  2022-09-27                      375.0   13.88   
5470617     Wyoming   Shop99  2022-09-28                      206.0  103.19   
5470618     Wyoming   Shop99  2022-09-29                      851.0  248.04   
5470619     Wyoming   Shop99  2022-09-30                      331.0   29.15   

Value_Type   revenue  
0           10378.80  
1            6346.85  
2            1204.84  
3            6899.45  
4            2163.06  
...              ...  
5470615      1488.04  
5470616       816.38  
5470617      2788.83  
5470618      5511.93  
5470619      1387.88  

[5470620 rows x 6 columns]

10. Compare the sets of `.stack()` and `.unstack()` with the set of `.melt()` and `.pivot()` from two perspectives: runtime and ease of use.

**Answer:** 

### Runtime
- **.stack() and .unstack():**
  - **Runtime Efficiency:** Generally, `.stack()` and `.unstack()` are faster for reshaping operations because they are specifically designed for handling multi-level indices in pandas. These functions are optimized for performance and can handle large datasets more efficiently.
  - **Example:** Using `.unstack()` to reshape data is typically quicker than using a combination of `.melt()` and `.pivot()` because it involves fewer intermediate steps and leverages internal optimizations.

- **.melt() and .pivot():**
  - **Runtime Efficiency:** The `.melt()` and `.pivot()` functions may introduce additional overhead due to the extra steps involved in reshaping the data. While they are still quite efficient, they may not perform as well as `.stack()` and `.unstack()` on very large datasets.
  - **Example:** Converting wide format data to long format with `.melt()` and then back to wide format with `.pivot()` can be slower because it involves creating and managing additional DataFrame objects.

### Ease of Use
- **.stack() and .unstack():**
  - **Ease of Use:** These functions are straightforward when dealing with multi-level indices. They are highly intuitive for hierarchical data structures and provide a simpler interface for common reshaping tasks.
  - **Example:** If you have a DataFrame with a multi-level index, `.unstack()` can easily convert it to a wide format and `.stack()` can revert it back. This simplicity and direct approach make them easier to understand and use effectively.

- **.melt() and .pivot():**
  - **Ease of Use:** While flexible, these functions can be more complex to use, especially for those unfamiliar with the details of data reshaping. Managing the additional steps and understanding the interactions between `.melt()` and `.pivot()` can be more challenging.
  - **Example:** Using `.melt()` to convert wide format data to long format and then `.pivot()` to revert it can require a deeper understanding of the DataFrame structure and more careful handling of column names and indices.

### Conclusion
- **Runtime:** `.stack()` and `.unstack()` are generally faster and more efficient for operations on multi-level indices, making them ideal for large datasets with hierarchical structures.
- **Ease of Use:** `.stack()` and `.unstack()` are easier to use for common reshaping tasks involving multi-level indices, providing a more straightforward and intuitive interface.

Overall, `.stack()` and `.unstack()` are preferred for both runtime efficiency and ease of use, making them superior for reshaping tasks in pandas.